File: minitensor/ops.py
import numpy as np
import cupy as cp

from .tensor import Tensor

import minitensor_cpp
import minitensor_cuda
import minitensor_cudnn

cudnn_handle = minitensor_cudnn.CuDNNHandleWrapper()

def exp(tensor):
    if tensor.device == "cpu":
        return Tensor(minitensor_cpp.exp(tensor._data))
    else:
        res = cp.empty_like(tensor._data)
        minitensor_cuda.exp(tensor._data.data.ptr, res.data.ptr, res.size)
        return Tensor(res)

def sum(tensor):
    if tensor.device == "cpu":
        return minitensor_cpp.sum(tensor._data)
    else:
        return minitensor_cudnn.sum(tensor._data.data.ptr, tensor._data.size, cudnn_handle)

File: minitensor/_C/cpp/ops.cpp
#include <cmath>
#include "ops.h"

namespace ops {

void exp(const float *in, float *out, const int size) {
  for (int i = 0; i < size; i++) {
    out[i] = std::exp(in[i]);
  }
}

void relu(const float *in, float *out, const int size) {
  for (int i = 0; i < size; i++) {
    out[i] = (in[i] > 0.0f) ? in[i] : 0.0f;
  }
}

float sum(const float *in, const int size) {
  float s = 0.0f;
  for (int i = 0; i < size; i++) {
    s += in[i];
  }
  return s;
}

} // namespace ops
